# CodeTune v2 — HumanEval Evaluation
Self-contained. No file uploads needed.

**Models**: `base` · `sft_A` · `sft_C`

## 0 · Configuration — fill in before running

In [ ]:
# ── EDIT THESE ──────────────────────────────────────────────────────────────
HF_TOKEN = "YOUR_HF_TOKEN"

BASE_MODEL  = "unsloth/Qwen3.5-9B"
SFT_A_MODEL = "Michlitt/codetune-v2-sft-A"
SFT_C_MODEL = "Michlitt/codetune-v2-sft-C"
DPO_MODEL   = "Michlitt/codetune-v2-dpo"   # dpo_v2 uploaded here

SFT_IS_ADAPTER = True

MAX_SEQ_LEN    = 2048
MAX_NEW_TOKENS = 512
LOAD_IN_4BIT   = True   # L4 22GB — bf16 Qwen3.5-9B ~18GB too tight

EXEC_TIMEOUT = 10
SAVE_EVERY   = 10

DRIVE_RESULTS = "/content/drive/MyDrive/codetune/eval_results"
# ────────────────────────────────────────────────────────────────────────────

MODELS = [
    (BASE_MODEL,  "base",   False),
    (SFT_A_MODEL, "sft_A",  SFT_IS_ADAPTER),
    (SFT_C_MODEL, "sft_C",  SFT_IS_ADAPTER),
    (DPO_MODEL,   "dpo_v1", SFT_IS_ADAPTER),   # old checkpoint — skip if already done
    (DPO_MODEL,   "dpo_v2", SFT_IS_ADAPTER),   # new 450-pair checkpoint
]

print("Models to evaluate:")
for m, lbl, adpt in MODELS:
    print(f"  [{lbl}]  {m}  {'(adapter)' if adpt else '(full)'}")
print(f"\nLOAD_IN_4BIT = {LOAD_IN_4BIT}")

## 1 · GPU check + install

In [2]:
import subprocess, torch
gpu = subprocess.run(
    ["nvidia-smi", "--query-gpu=name,memory.total,memory.free", "--format=csv,noheader"],
    capture_output=True, text=True
).stdout.strip()
print("GPU  :", gpu)
print("Torch:", torch.__version__, "| CUDA:", torch.version.cuda)
print("BF16 :", torch.cuda.is_bf16_supported())

GPU  : NVIDIA L4, 23034 MiB, 22564 MiB
Torch: 2.10.0+cu128 | CUDA: 12.8
BF16 : True


In [3]:
import os
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"
os.environ["WANDB_DISABLED"]          = "true"
os.environ["TOKENIZERS_PARALLELISM"]  = "false"

!pip install "unsloth[colab-new] @ git+https://github.com/unslothai/unsloth.git" -q
!pip install peft accelerate bitsandbytes datasets huggingface-hub tqdm -q
print("Done.")

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 401.6/401.6 kB 38.5 MB/s eta 0:00:00
Done.


## 2 · HF login + Drive mount

In [4]:
import os
from huggingface_hub import login

assert HF_TOKEN, "Set HF_TOKEN in the Config cell!"
login(token=HF_TOKEN, add_to_git_credential=False)
print("HF login OK")

try:
    from google.colab import drive
    drive.mount("/content/drive")
    os.makedirs(DRIVE_RESULTS, exist_ok=True)
    print("Drive mounted →", DRIVE_RESULTS)
except Exception as e:
    DRIVE_RESULTS = "/content/eval_results"
    os.makedirs(DRIVE_RESULTS, exist_ok=True)
    print(f"Drive unavailable ({e}). Saving to {DRIVE_RESULTS}")

HF login OK
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Drive mounted → /content/drive/MyDrive/codetune/eval_results


## 3 · Core functions

In [ ]:
from __future__ import annotations
import gc, json, re, subprocess, sys, tempfile
from pathlib import Path
import torch
from tqdm import tqdm

SYSTEM_PROMPT = (
    "You are an expert Python programmer. Complete the given function. "
    "Write only the function body — no extra explanation, no markdown fences, "
    "no test code."
)

TARGETED_TASKS = {
    "HumanEval/54":  "same_chars",
    "HumanEval/55":  "fib",
    "HumanEval/107": "minSubArraySum",
    "HumanEval/108": "intersection",
    "HumanEval/110": "prod_signs",
    "HumanEval/128": "tri",
    "HumanEval/141": "file_name_check",
    "HumanEval/147": "get_max_triples",
}


# ── Completion cleanup ────────────────────────────────────────────────
def clean_completion(raw: str, prompt: str) -> str:
    # Strip only trailing whitespace; preserve leading indent on first line
    text = raw.rstrip()

    # Strip Qwen3 think blocks
    text = re.sub(r"<think>[\s\S]*?</think>\s*", "", text)
    # Strip markdown fences
    text = re.sub(r"^```(?:python)?\s*\n", "", text)
    text = re.sub(r"\n?```\s*$", "", text)
    text = text.strip("\n")  # only strip newlines, not spaces

    # If model echoed the function header + docstring, strip them
    fn_match = re.search(r"^def\s+(\w+)", prompt, re.MULTILINE)
    if fn_match:
        fn = re.escape(fn_match.group(1))
        hdr = re.search(rf"def\s+{fn}\s*\(.*?\).*?:\s*\n", text, re.DOTALL)
        if hdr:
            after = text[hdr.end():]
            doc = re.match(r'\s*"""[\s\S]*?"""\s*\n', after)
            text = after[doc.end():] if doc else after
    text = text.strip("\n")

    # Normalize indentation to 4-space base.
    # Compute min indent across all non-empty lines (preserving leading spaces),
    # then shift so that the base becomes exactly 4.
    lines = text.splitlines()
    non_empty = [l for l in lines if l.strip()]
    if not non_empty:
        return text
    min_indent = min(len(l) - len(l.lstrip()) for l in non_empty)
    if min_indent != 4:
        shift = 4 - min_indent
        result = []
        for l in lines:
            if not l.strip():
                result.append("")
            elif shift > 0:
                result.append(" " * shift + l)
            else:
                result.append(l[-shift:])
        text = "\n".join(result)

    return text


# ── Model loading ─────────────────────────────────────────────────────
def load_model(model_id: str, is_adapter: bool):
    from unsloth import FastLanguageModel
    dtype = torch.bfloat16 if not LOAD_IN_4BIT else None
    if is_adapter:
        print(f"  Base  : {BASE_MODEL}")
        print(f"  Adapter: {model_id}")
        model, tokenizer = FastLanguageModel.from_pretrained(
            model_name=model_id,
            max_seq_length=MAX_SEQ_LEN,
            load_in_4bit=LOAD_IN_4BIT, dtype=dtype,
        )
    else:
        print(f"  Model: {model_id}")
        model, tokenizer = FastLanguageModel.from_pretrained(
            model_name=model_id, max_seq_length=MAX_SEQ_LEN,
            load_in_4bit=LOAD_IN_4BIT, dtype=dtype,
        )
    FastLanguageModel.for_inference(model)
    vram = torch.cuda.memory_allocated() / 1e9
    print(f"  Model loaded. VRAM: {vram:.1f} GB")
    return model, tokenizer

def unload_model(model, tokenizer):
    del model, tokenizer
    gc.collect()
    torch.cuda.empty_cache()
    vram = torch.cuda.memory_allocated() / 1e9
    print(f"  Model unloaded. VRAM in use: {vram:.1f} GB")


# ── Generation ────────────────────────────────────────────────────────
DEBUG_RAW = False

def generate(model, tokenizer, prompt: str, _debug_tid: str = "") -> str:
    _tok = getattr(tokenizer, "tokenizer", tokenizer)
    messages = [
        {"role": "system", "content": SYSTEM_PROMPT},
        {"role": "user",   "content": f"Complete this Python function:\n\n{prompt}"},
    ]
    text = _tok.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=True,
    )
    text = text + "<think>\n\n</think>\n\n"

    inputs = _tok(
        text=text, return_tensors="pt", truncation=True, max_length=MAX_SEQ_LEN,
    ).to(model.device)

    stop_ids = [_tok.eos_token_id]
    im_end = _tok.convert_tokens_to_ids("<|im_end|>")
    if isinstance(im_end, int) and im_end > 0:
        stop_ids.append(im_end)

    with torch.no_grad():
        out = model.generate(
            **inputs,
            max_new_tokens=MAX_NEW_TOKENS,
            do_sample=False,
            temperature=1.0,
            eos_token_id=stop_ids,
            pad_token_id=_tok.eos_token_id,
        )

    raw = _tok.decode(
        out[0][inputs["input_ids"].shape[1]:], skip_special_tokens=True
    ).strip()

    if DEBUG_RAW and _debug_tid:
        print(f"\n[DEBUG raw {_debug_tid}]: {raw!r}\n")

    return clean_completion(raw, prompt)


# ── Test execution ────────────────────────────────────────────────────
def run_tests(prompt: str, completion: str, test_code: str, entry_point: str) -> dict:
    full = prompt + completion + "\n\n" + test_code + f"\n\ncheck({entry_point})\n"
    with tempfile.NamedTemporaryFile(
        mode="w", suffix=".py", delete=False, encoding="utf-8"
    ) as f:
        f.write(full); tmp = f.name
    try:
        r = subprocess.run(
            [sys.executable, tmp], capture_output=True, text=True, timeout=EXEC_TIMEOUT,
        )
        return {"passed": r.returncode == 0, "stderr": r.stderr[:400]}
    except subprocess.TimeoutExpired:
        return {"passed": False, "stderr": "TIMEOUT"}
    except Exception as e:
        return {"passed": False, "stderr": str(e)}
    finally:
        Path(tmp).unlink(missing_ok=True)


# ── Checkpoint save ───────────────────────────────────────────────────
def _save_checkpoint(label: str, total: int, per_problem: dict, out_path: Path) -> dict:
    passed = sum(1 for v in per_problem.values() if v["passed"])
    summary = {
        "label": label, "total": total, "passed": passed,
        "pass_at_1": passed / total if total else 0.0,
        "failed_ids": [tid for tid, v in per_problem.items() if not v["passed"]],
        "per_problem": per_problem,
    }
    out_path.write_text(json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8")
    return summary


# ── Full evaluation ───────────────────────────────────────────────────
def evaluate(model, tokenizer, label: str, problems: list) -> dict:
    out_path = Path(DRIVE_RESULTS) / f"humaneval_{label}.json"

    per_problem = {}
    if out_path.exists():
        try:
            per_problem = json.loads(out_path.read_text(encoding="utf-8")).get("per_problem", {})
            print(f"  Resuming: {len(per_problem)}/{len(problems)} already done.")
        except Exception:
            pass

    remaining = [p for p in problems if p["task_id"] not in per_problem]
    total_done_start = len(per_problem)

    for i, prob in enumerate(tqdm(remaining, desc=f"[{label}]", unit="prob")):
        task_id    = prob["task_id"]
        debug_tid  = task_id if (DEBUG_RAW and i < 3) else ""
        completion = generate(model, tokenizer, prob["prompt"], _debug_tid=debug_tid)
        tr         = run_tests(prob["prompt"], completion, prob["test"], prob["entry_point"])

        per_problem[task_id] = {
            "passed":     tr["passed"],
            "completion": completion[:700],
            "stderr":     tr["stderr"],
        }

        n_done = total_done_start + i + 1
        flag   = "✓" if tr["passed"] else "✗"
        note   = " ← targeted" if task_id in TARGETED_TASKS else ""
        err    = f"  {tr['stderr'][:80]}" if not tr["passed"] and tr["stderr"] else ""
        print(f"  [{n_done:>3}/{len(problems)}] {flag} {task_id}{note}{err}")

        if (i + 1) % SAVE_EVERY == 0:
            _save_checkpoint(label, len(problems), per_problem, out_path)
            passed_so_far = sum(1 for v in per_problem.values() if v["passed"])
            print(f"    [checkpoint — {passed_so_far}/{n_done} passed]")

    summary = _save_checkpoint(label, len(problems), per_problem, out_path)
    p, t = summary["passed"], summary["total"]
    print(f"\n  ► {label}: {p}/{t} = {p/t:.1%}  pass@1")
    print(f"  ► Failed: {summary['failed_ids']}\n")
    return summary


print("Core functions loaded.")

## 4 · Load HumanEval dataset

In [6]:
from datasets import load_dataset
he_ds = load_dataset("openai/openai_humaneval", split="test", trust_remote_code=True)
HE_PROBLEMS = list(he_ds)
print(f"Loaded {len(HE_PROBLEMS)} problems.")

`trust_remote_code` is not supported anymore.
Please check that the Hugging Face dataset 'openai/openai_humaneval' isn't based on a loading script and remove `trust_remote_code`.
If the dataset is based on a loading script, please ask the dataset author to remove it and convert it to a standard format like Parquet.
ERROR:datasets.load:`trust_remote_code` is not supported anymore.
Please check that the Hugging Face dataset 'openai/openai_humaneval' isn't based on a loading script and remove `trust_remote_code`.
If the dataset is based on a loading script, please ask the dataset author to remove it and convert it to a standard format like Parquet.
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be 

Loaded 164 problems.


## 5 · Sanity check — 1 problem with sft_C
Confirm generation and test execution work before running all 164.

In [7]:
# ── accelerate 兼容补丁（Transformers 5.x _no_split_modules 改成了 set）──
import accelerate.utils.modeling as _am

_orig_gbm = _am.get_balanced_memory

def _fixed_gbm(model, max_memory=None, no_split_module_classes=None, **kwargs):
    if no_split_module_classes is not None:
        flat = []
        for c in no_split_module_classes:
            if isinstance(c, (set, frozenset)):
                flat.extend(str(x) for x in c)
            elif isinstance(c, type):
                flat.append(c.__name__)
            else:
                flat.append(str(c))
        no_split_module_classes = flat
    return _orig_gbm(
        model,
        max_memory=max_memory,
        no_split_module_classes=no_split_module_classes,
        **kwargs,
    )

_am.get_balanced_memory = _fixed_gbm
print("accelerate patched ✓")

accelerate patched ✓


In [8]:
print("Loading sft_C for sanity check...")
_m, _tok = load_model(SFT_C_MODEL, is_adapter=SFT_IS_ADAPTER)

prob = HE_PROBLEMS[0]
comp = generate(_m, _tok, prob["prompt"])
res  = run_tests(prob["prompt"], comp, prob["test"], prob["entry_point"])

print("=== Prompt ===")
print(prob["prompt"])
print("\n=== Completion ===")
print(comp)
print(f"\n=== Result: {'PASS ✓' if res['passed'] else 'FAIL ✗'} ===")
if not res["passed"]:
    print("stderr:", res["stderr"])

# Free memory before main eval
unload_model(_m, _tok)
del _m, _tok

Loading sft_C for sanity check...
🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
  Base  : unsloth/Qwen3.5-9B
  Adapter: Michlitt/codetune-v2-sft-C
==((====))==  Unsloth 2026.3.5: Fast Qwen3_5 patching. Transformers: 5.2.0.
   \\   /|    NVIDIA L4. Num GPUs = 1. Max memory: 22.034 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.10.0+cu128. CUDA: 8.9. CUDA Toolkit: 12.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: QLoRA and full finetuning all not selected. Switching to 16bit LoRA.


The fast path is not available because one of the required library is not installed. Falling back to torch implementation. To install follow https://github.com/fla-org/flash-linear-attention#installation and https://github.com/Dao-AILab/causal-conv1d


Loading weights:   0%|          | 0/760 [00:00<?, ?it/s]

=== Prompt ===
from typing import List


def has_close_elements(numbers: List[float], threshold: float) -> bool:
    """ Check if in given list of numbers, are any two numbers closer to each other than
    given threshold.
    >>> has_close_elements([1.0, 2.0, 3.0], 0.5)
    False
    >>> has_close_elements([1.0, 2.8, 3.0, 4.0, 5.0, 2.0], 0.3)
    True
    """


=== Completion ===
    numbers.sort()
    for i in range(len(numbers) - 1):
        if abs(numbers[i] - numbers[i + 1]) < threshold:
            return True
    return False

=== Result: PASS ✓ ===
  Model unloaded. VRAM in use: 19.3 GB


## 6 · Main evaluation — base · sft_A · sft_C
Loads each model in turn, runs 164 HumanEval problems, unloads before next model.
Saves checkpoint every 10 problems. Safe to re-run if interrupted — resumes from checkpoint.

In [12]:
all_summaries = {}

for model_id, label, is_adapter in MODELS:
    print(f"\n{'='*60}")
    print(f"Evaluating: {label}")
    print(f"{'='*60}")

    model, tokenizer = load_model(model_id, is_adapter)
    all_summaries[label] = evaluate(model, tokenizer, label, HE_PROBLEMS)
    unload_model(model, tokenizer)
    del model, tokenizer

print("\nAll evaluations complete.")


Evaluating: sft_C
  Base  : unsloth/Qwen3.5-9B
  Adapter: Michlitt/codetune-v2-sft-C
==((====))==  Unsloth 2026.3.5: Fast Qwen3_5 patching. Transformers: 5.2.0.
   \\   /|    NVIDIA L4. Num GPUs = 1. Max memory: 22.034 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.10.0+cu128. CUDA: 8.9. CUDA Toolkit: 12.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: QLoRA and full finetuning all not selected. Switching to 16bit LoRA.


Loading weights:   0%|          | 0/760 [00:00<?, ?it/s]

[sft_C]:   1%|          | 1/164 [00:07<21:40,  7.98s/prob]

  [  1/164] ✓ HumanEval/0


[sft_C]:   1%|          | 2/164 [00:21<30:33, 11.32s/prob]

  [  2/164] ✓ HumanEval/1


[sft_C]:   2%|▏         | 3/164 [00:24<19:53,  7.41s/prob]

  [  3/164] ✓ HumanEval/2


[sft_C]:   2%|▏         | 4/164 [00:30<18:34,  6.97s/prob]

  [  4/164] ✓ HumanEval/3


[sft_C]:   3%|▎         | 5/164 [00:37<17:58,  6.78s/prob]

  [  5/164] ✓ HumanEval/4


[sft_C]:   4%|▎         | 6/164 [00:43<17:46,  6.75s/prob]

  [  6/164] ✓ HumanEval/5


[sft_C]:   4%|▍         | 7/164 [00:57<23:31,  8.99s/prob]

  [  7/164] ✓ HumanEval/6


[sft_C]:   5%|▍         | 8/164 [01:01<19:39,  7.56s/prob]

  [  8/164] ✓ HumanEval/7


[sft_C]:   5%|▌         | 9/164 [01:08<19:05,  7.39s/prob]

  [  9/164] ✓ HumanEval/8


[sft_C]:   6%|▌         | 10/164 [01:17<19:58,  7.78s/prob]

  [ 10/164] ✗ HumanEval/9  Traceback (most recent call last):
  File "/tmp/tmp145kyuxv.py", line 33, in <mo
    [checkpoint — 9/10 passed]


[sft_C]:   7%|▋         | 11/164 [01:41<32:24, 12.71s/prob]

  [ 11/164] ✗ HumanEval/10  Traceback (most recent call last):
  File "/tmp/tmp2z3q6irh.py", line 38, in <mo


[sft_C]:   7%|▋         | 12/164 [01:46<26:16, 10.37s/prob]

  [ 12/164] ✓ HumanEval/11


[sft_C]:   8%|▊         | 13/164 [01:50<21:28,  8.53s/prob]

  [ 13/164] ✓ HumanEval/12


[sft_C]:   9%|▊         | 14/164 [01:56<18:53,  7.56s/prob]

  [ 14/164] ✓ HumanEval/13


[sft_C]:   9%|▉         | 15/164 [02:00<16:43,  6.73s/prob]

  [ 15/164] ✓ HumanEval/14


[sft_C]:  10%|▉         | 16/164 [02:04<14:31,  5.89s/prob]

  [ 16/164] ✓ HumanEval/15


[sft_C]:  10%|█         | 17/164 [02:07<12:17,  5.02s/prob]

  [ 17/164] ✓ HumanEval/16


[sft_C]:  11%|█         | 18/164 [02:18<16:40,  6.85s/prob]

  [ 18/164] ✓ HumanEval/17


[sft_C]:  12%|█▏        | 19/164 [02:27<17:39,  7.30s/prob]

  [ 19/164] ✓ HumanEval/18


[sft_C]:  12%|█▏        | 20/164 [02:45<25:15, 10.52s/prob]

  [ 20/164] ✓ HumanEval/19
    [checkpoint — 18/20 passed]


[sft_C]:  13%|█▎        | 21/164 [02:58<26:57, 11.31s/prob]

  [ 21/164] ✓ HumanEval/20


[sft_C]:  13%|█▎        | 22/164 [03:05<23:49, 10.06s/prob]

  [ 22/164] ✓ HumanEval/21


[sft_C]:  14%|█▍        | 23/164 [03:09<19:23,  8.25s/prob]

  [ 23/164] ✓ HumanEval/22


[sft_C]:  15%|█▍        | 24/164 [03:12<15:06,  6.47s/prob]

  [ 24/164] ✓ HumanEval/23


[sft_C]:  15%|█▌        | 25/164 [03:18<14:50,  6.41s/prob]

  [ 25/164] ✓ HumanEval/24


[sft_C]:  16%|█▌        | 26/164 [03:28<17:21,  7.55s/prob]

  [ 26/164] ✓ HumanEval/25


[sft_C]:  16%|█▋        | 27/164 [03:38<18:35,  8.15s/prob]

  [ 27/164] ✗ HumanEval/26  Traceback (most recent call last):
  File "/tmp/tmpiv2r602m.py", line 34, in <mo


[sft_C]:  17%|█▋        | 28/164 [03:40<14:41,  6.48s/prob]

  [ 28/164] ✓ HumanEval/27


[sft_C]:  18%|█▊        | 29/164 [03:43<11:56,  5.31s/prob]

  [ 29/164] ✓ HumanEval/28


[sft_C]:  18%|█▊        | 30/164 [03:47<11:14,  5.03s/prob]

  [ 30/164] ✓ HumanEval/29
    [checkpoint — 27/30 passed]


[sft_C]:  19%|█▉        | 31/164 [03:50<09:57,  4.49s/prob]

  [ 31/164] ✓ HumanEval/30


[sft_C]:  20%|█▉        | 32/164 [04:01<14:09,  6.43s/prob]

  [ 32/164] ✓ HumanEval/31


[sft_C]:  20%|██        | 33/164 [04:35<32:14, 14.76s/prob]

  [ 33/164] ✗ HumanEval/32  Traceback (most recent call last):
  File "/tmp/tmp8f3u3o_1.py", line 49, in <mo


[sft_C]:  21%|██        | 34/164 [04:54<34:21, 15.85s/prob]

  [ 34/164] ✓ HumanEval/33


[sft_C]:  21%|██▏       | 35/164 [04:56<25:15, 11.75s/prob]

  [ 35/164] ✓ HumanEval/34


[sft_C]:  22%|██▏       | 36/164 [04:58<18:54,  8.86s/prob]

  [ 36/164] ✓ HumanEval/35


[sft_C]:  23%|██▎       | 37/164 [05:06<18:04,  8.54s/prob]

  [ 37/164] ✓ HumanEval/36


[sft_C]:  23%|██▎       | 38/164 [05:21<21:48, 10.38s/prob]

  [ 38/164] ✓ HumanEval/37


[sft_C]:  24%|██▍       | 39/164 [05:38<26:15, 12.60s/prob]

  [ 39/164] ✗ HumanEval/38  Traceback (most recent call last):
  File "/tmp/tmpyggplzss.py", line 37, in <mo


[sft_C]:  24%|██▍       | 40/164 [05:59<30:57, 14.98s/prob]

  [ 40/164] ✓ HumanEval/39
    [checkpoint — 35/40 passed]


[sft_C]:  25%|██▌       | 41/164 [06:07<26:20, 12.85s/prob]

  [ 41/164] ✓ HumanEval/40


[sft_C]:  26%|██▌       | 42/164 [06:27<30:39, 15.08s/prob]

  [ 42/164] ✓ HumanEval/41


[sft_C]:  26%|██▌       | 43/164 [06:30<23:04, 11.44s/prob]

  [ 43/164] ✓ HumanEval/42


[sft_C]:  27%|██▋       | 44/164 [06:37<20:07, 10.06s/prob]

  [ 44/164] ✓ HumanEval/43


[sft_C]:  27%|██▋       | 45/164 [06:51<22:28, 11.33s/prob]

  [ 45/164] ✓ HumanEval/44


[sft_C]:  28%|██▊       | 46/164 [06:54<17:10,  8.74s/prob]

  [ 46/164] ✓ HumanEval/45


[sft_C]:  29%|██▊       | 47/164 [07:16<25:03, 12.85s/prob]

  [ 47/164] ✓ HumanEval/46


[sft_C]:  29%|██▉       | 48/164 [07:25<22:14, 11.50s/prob]

  [ 48/164] ✓ HumanEval/47


[sft_C]:  30%|██▉       | 49/164 [07:27<16:58,  8.86s/prob]

  [ 49/164] ✓ HumanEval/48


[sft_C]:  30%|███       | 50/164 [07:31<13:39,  7.19s/prob]

  [ 50/164] ✓ HumanEval/49
    [checkpoint — 45/50 passed]


[sft_C]:  31%|███       | 51/164 [07:39<14:19,  7.61s/prob]

  [ 51/164] ✗ HumanEval/50  Traceback (most recent call last):
  File "/tmp/tmp38my15kh.py", line 34, in <mo


[sft_C]:  32%|███▏      | 52/164 [07:44<12:44,  6.83s/prob]

  [ 52/164] ✓ HumanEval/51


[sft_C]:  32%|███▏      | 53/164 [07:48<10:43,  5.80s/prob]

  [ 53/164] ✓ HumanEval/52


[sft_C]:  33%|███▎      | 54/164 [07:51<09:01,  4.92s/prob]

  [ 54/164] ✓ HumanEval/53


[sft_C]:  34%|███▎      | 55/164 [07:54<08:15,  4.54s/prob]

  [ 55/164] ✗ HumanEval/54 ← targeted  Traceback (most recent call last):
  File "/tmp/tmp41ad87gb.py", line 37, in <mo


[sft_C]:  34%|███▍      | 56/164 [08:04<11:15,  6.25s/prob]

  [ 56/164] ✓ HumanEval/55 ← targeted


[sft_C]:  35%|███▍      | 57/164 [08:13<12:25,  6.97s/prob]

  [ 57/164] ✓ HumanEval/56


[sft_C]:  35%|███▌      | 58/164 [08:23<13:47,  7.80s/prob]

  [ 58/164] ✓ HumanEval/57


[sft_C]:  36%|███▌      | 59/164 [08:27<11:29,  6.57s/prob]

  [ 59/164] ✓ HumanEval/58


[sft_C]:  37%|███▋      | 60/164 [08:33<11:24,  6.59s/prob]

  [ 60/164] ✓ HumanEval/59
    [checkpoint — 53/60 passed]


[sft_C]:  37%|███▋      | 61/164 [08:36<09:34,  5.58s/prob]

  [ 61/164] ✓ HumanEval/60


[sft_C]:  38%|███▊      | 62/164 [08:45<11:08,  6.55s/prob]

  [ 62/164] ✓ HumanEval/61


[sft_C]:  38%|███▊      | 63/164 [08:49<09:33,  5.68s/prob]

  [ 63/164] ✓ HumanEval/62


[sft_C]:  39%|███▉      | 64/164 [09:02<13:19,  7.99s/prob]

  [ 64/164] ✓ HumanEval/63


[sft_C]:  40%|███▉      | 65/164 [10:08<41:37, 25.23s/prob]

  [ 65/164] ✗ HumanEval/64  Traceback (most recent call last):
  File "/tmp/tmp08t1aeyi.py", line 43, in <mo


[sft_C]:  40%|████      | 66/164 [10:17<33:19, 20.41s/prob]

  [ 66/164] ✓ HumanEval/65


[sft_C]:  41%|████      | 67/164 [10:20<24:35, 15.22s/prob]

  [ 67/164] ✓ HumanEval/66


[sft_C]:  41%|████▏     | 68/164 [10:26<19:43, 12.32s/prob]

  [ 68/164] ✗ HumanEval/67  Traceback (most recent call last):
  File "/tmp/tmp7qdaz59g.py", line 31, in <mo


[sft_C]:  42%|████▏     | 69/164 [10:38<19:28, 12.30s/prob]

  [ 69/164] ✓ HumanEval/68


[sft_C]:  43%|████▎     | 70/164 [10:49<18:36, 11.87s/prob]

  [ 70/164] ✓ HumanEval/69
    [checkpoint — 61/70 passed]


[sft_C]:  43%|████▎     | 71/164 [10:56<16:28, 10.63s/prob]

  [ 71/164] ✓ HumanEval/70


[sft_C]:  44%|████▍     | 72/164 [11:08<16:50, 10.98s/prob]

  [ 72/164] ✓ HumanEval/71


[sft_C]:  45%|████▍     | 73/164 [11:13<13:59,  9.23s/prob]

  [ 73/164] ✓ HumanEval/72


[sft_C]:  45%|████▌     | 74/164 [11:21<12:57,  8.64s/prob]

  [ 74/164] ✓ HumanEval/73


[sft_C]:  46%|████▌     | 75/164 [11:30<13:08,  8.86s/prob]

  [ 75/164] ✓ HumanEval/74


[sft_C]:  46%|████▋     | 76/164 [11:52<18:46, 12.80s/prob]

  [ 76/164] ✗ HumanEval/75  Traceback (most recent call last):
  File "/tmp/tmpc541_vub.py", line 47, in <mo


[sft_C]:  47%|████▋     | 77/164 [12:08<19:59, 13.79s/prob]

  [ 77/164] ✓ HumanEval/76


[sft_C]:  48%|████▊     | 78/164 [12:11<15:18, 10.68s/prob]

  [ 78/164] ✗ HumanEval/77  Traceback (most recent call last):
  File "/tmp/tmp3_b4pufq.py", line 34, in <mo


[sft_C]:  48%|████▊     | 79/164 [12:19<13:53,  9.80s/prob]

  [ 79/164] ✓ HumanEval/78


[sft_C]:  49%|████▉     | 80/164 [12:23<11:15,  8.04s/prob]

  [ 80/164] ✓ HumanEval/79
    [checkpoint — 69/80 passed]


[sft_C]:  49%|████▉     | 81/164 [12:31<10:56,  7.91s/prob]

  [ 81/164] ✓ HumanEval/80


[sft_C]:  50%|█████     | 82/164 [13:04<21:11, 15.50s/prob]

  [ 82/164] ✗ HumanEval/81  Traceback (most recent call last):
  File "/tmp/tmpm3j194c6.py", line 70, in <mo


[sft_C]:  51%|█████     | 83/164 [13:17<19:57, 14.79s/prob]

  [ 83/164] ✓ HumanEval/82


[sft_C]:  51%|█████     | 84/164 [13:25<16:50, 12.63s/prob]

  [ 84/164] ✗ HumanEval/83  Traceback (most recent call last):
  File "/tmp/tmpf87mo85j.py", line 27, in <mo


[sft_C]:  52%|█████▏    | 85/164 [13:48<20:46, 15.78s/prob]

  [ 85/164] ✓ HumanEval/84


[sft_C]:  52%|█████▏    | 86/164 [13:53<16:17, 12.54s/prob]

  [ 86/164] ✓ HumanEval/85


[sft_C]:  53%|█████▎    | 87/164 [14:01<14:21, 11.19s/prob]

  [ 87/164] ✗ HumanEval/86  Traceback (most recent call last):
  File "/tmp/tmpuw0lyh4h.py", line 39, in <mo


[sft_C]:  54%|█████▎    | 88/164 [14:21<17:39, 13.95s/prob]

  [ 88/164] ✓ HumanEval/87


[sft_C]:  54%|█████▍    | 89/164 [14:31<16:02, 12.83s/prob]

  [ 89/164] ✓ HumanEval/88


[sft_C]:  55%|█████▍    | 90/164 [14:46<16:22, 13.28s/prob]

  [ 90/164] ✓ HumanEval/89
    [checkpoint — 76/90 passed]


[sft_C]:  55%|█████▌    | 91/164 [14:53<14:00, 11.51s/prob]

  [ 91/164] ✓ HumanEval/90


[sft_C]:  56%|█████▌    | 92/164 [14:59<11:49,  9.85s/prob]

  [ 92/164] ✗ HumanEval/91  Traceback (most recent call last):
  File "/tmp/tmpejlzzqvo.py", line 33, in <mo


[sft_C]:  57%|█████▋    | 93/164 [15:07<10:56,  9.25s/prob]

  [ 93/164] ✓ HumanEval/92


[sft_C]:  57%|█████▋    | 94/164 [15:26<14:14, 12.20s/prob]

  [ 94/164] ✗ HumanEval/93  Traceback (most recent call last):
  File "/tmp/tmpde55npkk.py", line 44, in <mo


[sft_C]:  58%|█████▊    | 95/164 [15:44<16:01, 13.94s/prob]

  [ 95/164] ✓ HumanEval/94


[sft_C]:  59%|█████▊    | 96/164 [15:58<15:42, 13.87s/prob]

  [ 96/164] ✗ HumanEval/95  Traceback (most recent call last):
  File "/tmp/tmph6jad928.py", line 47, in <mo


[sft_C]:  59%|█████▉    | 97/164 [16:18<17:41, 15.84s/prob]

  [ 97/164] ✗ HumanEval/96  Traceback (most recent call last):
  File "/tmp/tmp23mtaq2b.py", line 44, in <mo


[sft_C]:  60%|█████▉    | 98/164 [16:22<13:19, 12.12s/prob]

  [ 98/164] ✓ HumanEval/97


[sft_C]:  60%|██████    | 99/164 [16:28<11:19, 10.45s/prob]

  [ 99/164] ✓ HumanEval/98


[sft_C]:  61%|██████    | 100/164 [16:34<09:47,  9.19s/prob]

  [100/164] ✓ HumanEval/99
    [checkpoint — 82/100 passed]


[sft_C]:  62%|██████▏   | 101/164 [16:45<10:04,  9.59s/prob]

  [101/164] ✓ HumanEval/100


[sft_C]:  62%|██████▏   | 102/164 [16:47<07:40,  7.42s/prob]

  [102/164] ✓ HumanEval/101


[sft_C]:  63%|██████▎   | 103/164 [16:53<07:03,  6.95s/prob]

  [103/164] ✓ HumanEval/102


[sft_C]:  63%|██████▎   | 104/164 [17:05<08:25,  8.43s/prob]

  [104/164] ✓ HumanEval/103


[sft_C]:  64%|██████▍   | 105/164 [17:12<07:45,  7.88s/prob]

  [105/164] ✓ HumanEval/104


[sft_C]:  65%|██████▍   | 106/164 [17:27<09:50, 10.17s/prob]

  [106/164] ✓ HumanEval/105


[sft_C]:  65%|██████▌   | 107/164 [17:39<10:03, 10.59s/prob]

  [107/164] ✗ HumanEval/106  Traceback (most recent call last):
  File "/tmp/tmphd2vf63k.py", line 29, in <mo


[sft_C]:  66%|██████▌   | 108/164 [17:52<10:41, 11.46s/prob]

  [108/164] ✓ HumanEval/107 ← targeted


[sft_C]:  66%|██████▋   | 109/164 [18:01<09:49, 10.72s/prob]

  [109/164] ✗ HumanEval/108 ← targeted  Traceback (most recent call last):
  File "/tmp/tmp1vgncxxl.py", line 38, in <mo


[sft_C]:  67%|██████▋   | 110/164 [18:07<08:17,  9.22s/prob]

  [110/164] ✗ HumanEval/109  Traceback (most recent call last):
  File "/tmp/tmpossxytag.py", line 45, in <mo
    [checkpoint — 89/110 passed]


[sft_C]:  68%|██████▊   | 111/164 [18:19<08:46,  9.93s/prob]

  [111/164] ✓ HumanEval/110 ← targeted


[sft_C]:  68%|██████▊   | 112/164 [18:34<10:01, 11.56s/prob]

  [112/164] ✓ HumanEval/111


[sft_C]:  69%|██████▉   | 113/164 [18:39<08:07,  9.56s/prob]

  [113/164] ✓ HumanEval/112


[sft_C]:  70%|██████▉   | 114/164 [18:52<08:50, 10.61s/prob]

  [114/164] ✓ HumanEval/113


[sft_C]:  70%|███████   | 115/164 [19:01<08:17, 10.15s/prob]

  [115/164] ✓ HumanEval/114


[sft_C]:  71%|███████   | 116/164 [19:06<06:51,  8.58s/prob]

  [116/164] ✗ HumanEval/115  Traceback (most recent call last):
  File "/tmp/tmpyii8pwzb.py", line 57, in <mo


[sft_C]:  71%|███████▏  | 117/164 [19:19<07:43,  9.87s/prob]

  [117/164] ✓ HumanEval/116


[sft_C]:  72%|███████▏  | 118/164 [19:28<07:23,  9.64s/prob]

  [118/164] ✓ HumanEval/117


[sft_C]:  73%|███████▎  | 119/164 [19:45<08:54, 11.87s/prob]

  [119/164] ✓ HumanEval/118


[sft_C]:  73%|███████▎  | 120/164 [20:00<09:29, 12.94s/prob]

  [120/164] ✓ HumanEval/119
    [checkpoint — 98/120 passed]


[sft_C]:  74%|███████▍  | 121/164 [20:03<07:03,  9.85s/prob]

  [121/164] ✗ HumanEval/120  Traceback (most recent call last):
  File "/tmp/tmpbf_90qbu.py", line 48, in <mo


[sft_C]:  74%|███████▍  | 122/164 [20:08<05:51,  8.37s/prob]

  [122/164] ✓ HumanEval/121


[sft_C]:  75%|███████▌  | 123/164 [20:14<05:14,  7.68s/prob]

  [123/164] ✓ HumanEval/122


[sft_C]:  76%|███████▌  | 124/164 [20:34<07:30, 11.26s/prob]

  [124/164] ✓ HumanEval/123


[sft_C]:  76%|███████▌  | 125/164 [20:55<09:15, 14.25s/prob]

  [125/164] ✓ HumanEval/124


[sft_C]:  77%|███████▋  | 126/164 [21:07<08:39, 13.67s/prob]

  [126/164] ✓ HumanEval/125


[sft_C]:  77%|███████▋  | 127/164 [22:12<17:52, 28.99s/prob]

  [127/164] ✗ HumanEval/126    File "/tmp/tmprxp7p5or.py", line 76
    if lst[26]
              ^
SyntaxError


[sft_C]:  78%|███████▊  | 128/164 [22:30<15:26, 25.75s/prob]

  [128/164] ✓ HumanEval/127


[sft_C]:  79%|███████▊  | 129/164 [22:39<12:06, 20.75s/prob]

  [129/164] ✗ HumanEval/128 ← targeted  Traceback (most recent call last):
  File "/tmp/tmp0_j9vtc4.py", line 42, in <mo


[sft_C]:  79%|███████▉  | 130/164 [22:49<09:56, 17.55s/prob]

  [130/164] ✗ HumanEval/129  Traceback (most recent call last):
  File "/tmp/tmpcseuz9sa.py", line 60, in <mo
    [checkpoint — 104/130 passed]


[sft_C]:  80%|███████▉  | 131/164 [23:05<09:17, 16.90s/prob]

  [131/164] ✗ HumanEval/130  Traceback (most recent call last):
  File "/tmp/tmp2myrino4.py", line 50, in <mo


[sft_C]:  80%|████████  | 132/164 [23:12<07:25, 13.92s/prob]

  [132/164] ✗ HumanEval/131  Traceback (most recent call last):
  File "/tmp/tmp8tt80_w9.py", line 31, in <mo


[sft_C]:  81%|████████  | 133/164 [23:20<06:21, 12.32s/prob]

  [133/164] ✗ HumanEval/132  Traceback (most recent call last):
  File "/tmp/tmpomun1xs8.py", line 49, in <mo


[sft_C]:  82%|████████▏ | 134/164 [23:24<04:51,  9.72s/prob]

  [134/164] ✗ HumanEval/133  Traceback (most recent call last):
  File "/tmp/tmpouujzwej.py", line 40, in <mo


[sft_C]:  82%|████████▏ | 135/164 [23:31<04:18,  8.91s/prob]

  [135/164] ✗ HumanEval/134  Traceback (most recent call last):
  File "/tmp/tmpg4wmqzr7.py", line 41, in <mo


[sft_C]:  83%|████████▎ | 136/164 [23:37<03:44,  8.02s/prob]

  [136/164] ✗ HumanEval/135  Traceback (most recent call last):
  File "/tmp/tmpjyc8ways.py", line 30, in <mo


[sft_C]:  84%|████████▎ | 137/164 [23:48<04:05,  9.10s/prob]

  [137/164] ✓ HumanEval/136


[sft_C]:  84%|████████▍ | 138/164 [23:58<04:00,  9.23s/prob]

  [138/164] ✗ HumanEval/137  Traceback (most recent call last):
  File "/tmp/tmp8xekgzsu.py", line 42, in <mo


[sft_C]:  85%|████████▍ | 139/164 [24:07<03:45,  9.03s/prob]

  [139/164] ✗ HumanEval/138  Traceback (most recent call last):
  File "/tmp/tmpged_8m1s.py", line 25, in <mo


[sft_C]:  85%|████████▌ | 140/164 [24:17<03:46,  9.45s/prob]

  [140/164] ✓ HumanEval/139
    [checkpoint — 106/140 passed]


[sft_C]:  86%|████████▌ | 141/164 [24:25<03:26,  8.97s/prob]

  [141/164] ✓ HumanEval/140


[sft_C]:  87%|████████▋ | 142/164 [24:39<03:50, 10.50s/prob]

  [142/164] ✓ HumanEval/141 ← targeted


[sft_C]:  87%|████████▋ | 143/164 [24:48<03:31, 10.07s/prob]

  [143/164] ✗ HumanEval/142  Traceback (most recent call last):
  File "/tmp/tmpypss4c56.py", line 43, in <mo


[sft_C]:  88%|████████▊ | 144/164 [25:05<04:01, 12.07s/prob]

  [144/164] ✓ HumanEval/143


[sft_C]:  88%|████████▊ | 145/164 [25:12<03:25, 10.80s/prob]

  [145/164] ✓ HumanEval/144


[sft_C]:  89%|████████▉ | 146/164 [25:19<02:52,  9.58s/prob]

  [146/164] ✗ HumanEval/145  Traceback (most recent call last):
  File "/tmp/tmpd63rcwc4.py", line 33, in <mo


[sft_C]:  90%|████████▉ | 147/164 [25:30<02:51, 10.06s/prob]

  [147/164] ✓ HumanEval/146


[sft_C]:  90%|█████████ | 148/164 [25:44<02:59, 11.25s/prob]

  [148/164] ✓ HumanEval/147 ← targeted


[sft_C]:  91%|█████████ | 149/164 [26:00<03:09, 12.63s/prob]

  [149/164] ✓ HumanEval/148


[sft_C]:  91%|█████████▏| 150/164 [26:09<02:42, 11.59s/prob]

  [150/164] ✓ HumanEval/149
    [checkpoint — 114/150 passed]


[sft_C]:  92%|█████████▏| 151/164 [26:20<02:26, 11.30s/prob]

  [151/164] ✓ HumanEval/150


[sft_C]:  93%|█████████▎| 152/164 [26:25<01:50,  9.25s/prob]

  [152/164] ✓ HumanEval/151


[sft_C]:  93%|█████████▎| 153/164 [26:28<01:22,  7.51s/prob]

  [153/164] ✓ HumanEval/152


[sft_C]:  94%|█████████▍| 154/164 [26:42<01:34,  9.42s/prob]

  [154/164] ✓ HumanEval/153


[sft_C]:  95%|█████████▍| 155/164 [26:48<01:16,  8.49s/prob]

  [155/164] ✓ HumanEval/154


[sft_C]:  95%|█████████▌| 156/164 [26:59<01:12,  9.07s/prob]

  [156/164] ✓ HumanEval/155


[sft_C]:  96%|█████████▌| 157/164 [27:24<01:38, 14.10s/prob]

  [157/164] ✓ HumanEval/156


[sft_C]:  96%|█████████▋| 158/164 [27:30<01:09, 11.66s/prob]

  [158/164] ✓ HumanEval/157


[sft_C]:  97%|█████████▋| 159/164 [27:47<01:05, 13.04s/prob]

  [159/164] ✓ HumanEval/158


[sft_C]:  98%|█████████▊| 160/164 [28:06<00:59, 14.80s/prob]

  [160/164] ✗ HumanEval/159  Traceback (most recent call last):
  File "/tmp/tmp2x_bz4e7.py", line 53, in <mo
    [checkpoint — 123/160 passed]


[sft_C]:  98%|█████████▊| 161/164 [28:22<00:45, 15.15s/prob]

  [161/164] ✗ HumanEval/160  Traceback (most recent call last):
  File "/tmp/tmp47xv1l0x.py", line 53, in <mo


[sft_C]:  99%|█████████▉| 162/164 [28:58<00:42, 21.39s/prob]

  [162/164] ✓ HumanEval/161


[sft_C]:  99%|█████████▉| 163/164 [29:04<00:16, 16.94s/prob]

  [163/164] ✓ HumanEval/162


[sft_C]: 100%|██████████| 164/164 [29:14<00:00, 10.70s/prob]

  [164/164] ✗ HumanEval/163  Traceback (most recent call last):
  File "/tmp/tmpup7f2lac.py", line 32, in <mo

  ► sft_C: 125/164 = 76.2%  pass@1
  ► Failed: ['HumanEval/9', 'HumanEval/10', 'HumanEval/26', 'HumanEval/32', 'HumanEval/38', 'HumanEval/50', 'HumanEval/54', 'HumanEval/64', 'HumanEval/67', 'HumanEval/75', 'HumanEval/77', 'HumanEval/81', 'HumanEval/83', 'HumanEval/86', 'HumanEval/91', 'HumanEval/93', 'HumanEval/95', 'HumanEval/96', 'HumanEval/106', 'HumanEval/108', 'HumanEval/109', 'HumanEval/115', 'HumanEval/120', 'HumanEval/126', 'HumanEval/128', 'HumanEval/129', 'HumanEval/130', 'HumanEval/131', 'HumanEval/132', 'HumanEval/133', 'HumanEval/134', 'HumanEval/135', 'HumanEval/137', 'HumanEval/138', 'HumanEval/142', 'HumanEval/145', 'HumanEval/159', 'HumanEval/160', 'HumanEval/163']



  Model unloaded. VRAM in use: 20.7 GB

All evaluations complete.


## 7 · Results summary

In [16]:
import json
from pathlib import Path

# Load all saved results (works even if cell 6 was split across sessions)
all_data = {}
for model_id, label, _ in MODELS:
    fp = Path(DRIVE_RESULTS) / f"humaneval_{label}.json"
    if fp.exists():
        all_data[label] = json.loads(fp.read_text(encoding="utf-8"))
    else:
        print(f"WARNING: {fp} not found — run cell 6 first.")

labels = [lbl for _, lbl, _ in MODELS if lbl in all_data]

# ── Overall pass@1 ───────────────────────────────────────────────────
print(f"\n{'Model':<10} {'pass@1':>8} {'passed':>12} {'failed':>8}")
print("-" * 42)
for lbl in labels:
    d = all_data[lbl]
    print(f"{lbl:<10} {d['pass_at_1']:>8.1%} "
          f"{d['passed']:>5}/{d['total']:<5} {d['total']-d['passed']:>6}")

# ── Targeted tasks ───────────────────────────────────────────────────
col = 8
w   = 28 + col * len(labels)
print(f"\n\nTargeted failure cases  (Coder-Agent baseline: 0/8)")
print("=" * w)
print(f"{'Task':<28}" + "".join(f"{l:>{col}}" for l in labels))
print("-" * w)
for tid, fn in TARGETED_TASKS.items():
    row = f"{tid + ' (' + fn + ')':<28}"
    for lbl in labels:
        v = all_data[lbl].get("per_problem", {}).get(tid, {}).get("passed", None)
        row += f"{'✓' if v else ('✗' if v is False else '?'):>{col}}"
    print(row)
print("-" * w)
totals = f"{'TARGETED TOTAL':<28}"
for lbl in labels:
    per = all_data[lbl].get("per_problem", {})
    n = sum(1 for tid in TARGETED_TASKS if per.get(tid, {}).get("passed"))
    totals += f"{f'{n}/8':>{col}}"
print(totals)
print("=" * w)


Model        pass@1       passed   failed
------------------------------------------
base          37.8%    62/164      102
sft_A         65.2%   107/164       57
sft_C         76.2%   125/164       39


Targeted failure cases  (Coder-Agent baseline: 0/8)
Task                            base   sft_A   sft_C
----------------------------------------------------
HumanEval/54 (same_chars)          ✗       ✗       ✗
HumanEval/55 (fib)                 ✗       ✓       ✓
HumanEval/107 (minSubArraySum)       ✓       ✓       ✓
HumanEval/108 (intersection)       ✗       ✗       ✗
HumanEval/110 (prod_signs)         ✗       ✗       ✓
HumanEval/128 (tri)                ✗       ✓       ✗
HumanEval/141 (file_name_check)       ✗       ✗       ✓
HumanEval/147 (get_max_triples)       ✓       ✓       ✓
----------------------------------------------------
TARGETED TOTAL                   2/8     4/8     5/8


## 8 · (Debug) Inspect specific failures

In [18]:
INSPECT_LABEL   = "sft_C"          # base / sft_A / sft_C
INSPECT_TASK_ID = None   # or None to show first N failures
MAX_SHOW        = 8

fp = Path(DRIVE_RESULTS) / f"humaneval_{INSPECT_LABEL}.json"
d  = json.loads(fp.read_text(encoding="utf-8"))
per = d["per_problem"]

tasks = [INSPECT_TASK_ID] if INSPECT_TASK_ID else [
    tid for tid, v in per.items() if not v["passed"]
][:MAX_SHOW]

for tid in tasks:
    v = per.get(tid, {})
    print(f"\n{'='*60}")
    print(f"{tid}  —  {'PASS ✓' if v.get('passed') else 'FAIL ✗'}")
    print(f"Completion:\n{v.get('completion', '(none)')}")
    if not v.get("passed"):
        print(f"stderr: {v.get('stderr', '')}")


HumanEval/9  —  FAIL ✗
Completion:
    max_num = numbers[0]
    result = [max_num]
    for num in numbers[1:]:
        if num > max_num:
            max_num = num
        result.append(max_num)
    return result
stderr: Traceback (most recent call last):
  File "/tmp/tmp145kyuxv.py", line 33, in <module>
    check(rolling_max)
  File "/tmp/tmp145kyuxv.py", line 27, in check
    assert candidate([]) == []
           ^^^^^^^^^^^^^
  File "/tmp/tmp145kyuxv.py", line 10, in rolling_max
    max_num = numbers[0]
              ~~~~~~~^^^
IndexError: list index out of range


HumanEval/10  —  FAIL ✗
Completion:
    return string == string[::-1]
stderr: Traceback (most recent call last):
  File "/tmp/tmp2z3q6irh.py", line 38, in <module>
    check(make_palindrome)
  File "/tmp/tmp2z3q6irh.py", line 31, in check
    assert candidate('') == ''
           ^^^^^^^^^^^^^^^^^^^
AssertionError


HumanEval/26  —  FAIL ✗
Completion:
    count_dict = {}
    result = []
    for num in numbers:
        if